V1

devusexu opened the pull request.

devusexu committed:
feat: implement user registration and login

Configure the project to use the 'jwt-auth' package for user authentication. This includes setting up registration and login functionality for users.

devusexu committed:
refactor: customize 404 response handling

devusexu committed:
chore: define migrations and relationships between models

devusexu committed:
chore: add column 'proposed_username' to token migration

devusexu committed:
feat: implement student registration using unique tokens

Mentors can generate and manage registration tokens for students. The token usage is tracked, and upon use, the token is invalidated and user details are recorded.

devusexu committed:
feat: implement student progress monitoring for admins

1. resources/views/progress.blade.php => admins could view the progress of all students
2. routes/web.php => add a route to view progress and a route to update 'start_date' and 'leave_date'
3. app/Models/User.php => add custom attributes which are derived from other attributes

JYu1999 commented on app/Models/RegistrationToken.php:
```diff
@@ -0,0 +1,18 @@
+{
+    use HasFactory;
+
+    protected $guarded = [];
```
盡量避免使用 [Mass Assignment](https://laravel.com/docs/10.x/eloquent#mass-assignment)

JYu1999 commented on app/Http/Controllers/RegistrationTokenController.php:
```diff
@@ -0,0 +1,42 @@
+           $token = Str::random(10);
+       } while (RegistrationToken::where('value', $token)->where('is_valid', 1)->exists());
+
+       // The link should be updated with real URL after deployment
+       $registrationLink = 'http://localhost:3000/client?token=' . $token;
```
在正式站要改網址的話，要如何確保不會忘記改呢？
或是如果有多處同時使用到此網址怎麼辦？

JYu1999 commented on app/Http/Controllers/RegisterController.php:
```diff
@@ -0,0 +1,48 @@
+        $user = User::create($validated);
+
+        // Update registration_tokens
+        $record->is_valid = 0;
```
我想問為什麼這裡是用 snake case？

JYu1999 commented on routes/web.php:
```diff
@@ -16,3 +19,36 @@
+Route::get('/tokens', [RegistrationTokenController::class, 'index']);
+Route::post('/tokens', [RegistrationTokenController::class, 'store']);
+
+Route::get('/students', function () {
+    return view('progress', [
+        'students' => User::where('role', 'student')->get()
+    ]);
+});
+
+Route::patch('/students/{user}', function (Request $request, User $user) {
+    $validated = $request->validate([
+        'start_date' => ['date'],
+        'leave_date' => ['date'],
+    ]);
+
+    // start_date could be set only if it's null and request body contains start_date
+    if (is_null($user->start_date) && isset($validated['start_date'])) {
+        $user->start_date = $validated['start_date'];
+
+        // proposed_leave_date is six months after start_date
+        $user->proposed_leave_date = (new DateTime($validated['start_date']))->modify('+6 months');
+    }
+
+    // leave_date could be set only if it's null and request body contains leave_date
+    if (is_null($user->leave_date) && isset($validated['leave_date'])) {
+        $user->leave_date = $validated['leave_date'];
+    }
+
+    $user->save();
+
+    return redirect('/students');
+});
```
基本上不會在 Router 裡面寫邏輯

devusexu replied:
已依照上述三項問題更改

devusexu committed:
refactor: replace hard-coded URL with configuration setting

devusexu committed:
refactor: switch from $guarded to $fillable for explicit attribute definition

devusexu committed:
refactor: move route logic to 'StudentController'

JYu1999 commented on app/Http/Controllers/RegisterController.php:
```diff
@@ -0,0 +1,48 @@
+           return response('Your registration token is incorrect', 422);
+        }
+
+        $user = User::create($validated);
```
我發現我留言留錯行了，有沒有用到 Mass Assignment 的重點不在於 guarded 或 filled array 裡面存了什麼，是在於你實際上存資料的時候怎麼存

JYu1999 replied:
還是你認為沒有必要改這件事情，我想聽一下你的看法？

devusexu replied:
是因為 $request 可能帶有更多不用受到檢驗的 field 嗎？我想說不在 $fillable 裡面的 field 會被自動省略，所以手動 assign 那四項與一次 assign 同樣意思
這樣可能會有哪些危險？

JYu1999 replied:
我覺得或許可以放大到維護面來思考
如果使用 `$fillable[]` or `$guarded[]` 一定是可以達到一樣的效果
但之後每次新增新的欄位，可能都需要考慮它有沒有需要被放進 Array
甚至有可能明明是敏感欄位，但卻忘記放進 `$guarded[]` 裡面
以這個角度來看，好像全面禁止 mass assignment 是對後續維護比較友善的作法？

devusexu replied:
你是說像這樣寫比較好嗎？所有欄位都分開 assign
``` php
$registrationToken = new RegistrationToken();
$registrationToken->proposed_username = $validated['proposed_username'];
$registrationToken->value = $token;
$registrationToken->expired_time = (new DateTime())->modify('+30 days')->format('Y-m-d H:i:s');
$registrationToken->save();
```

JYu1999 replied:
確實我的意思是所有欄位分開 assign
但我沒有覺得怎麼樣一定比較好，你可以自己評估看看～

JYu1999 commented on app/Http/Controllers/RegistrationTokenController.php:
```diff
@@ -0,0 +1,42 @@
+
+    public function store(Request $request)
+    {
+       $validated = $request->validate([
+           'proposed_username' => ['required', 'string', 'max:50'],
+       ]);
+
+       // Generate a token, and if a token with the same value already exists, generate a new one.
+       do {
+           $token = Str::random(10);
+       } while (RegistrationToken::where('value', $token)->where('is_valid', 1)->exists());
+
+       // The link should be updated with real URL after deployment
+       $registrationLink = config('app.frontend_url') . '?token=' . $token;
+
+       RegistrationToken::create(
+           array_merge($validated, [
+               'value' => $token,
```
是不是可以評估下，哪些功能可以放進 service 裡面實作，維持 Controller 單純性？
還是應該放在 Laravel 就好？

devusexu replied:
如果是你做這種 project 的話，你會建議怎麼樣的結構？例如驗證輸入要不要創單獨的 form request，service 裡面要不要再把與資料庫互動的部份獨立成 repository

我也擔心如果專案太小的話會不會有點做過頭

JYu1999 replied:
欸不是你怎麼把問題丟回來給我XD

JYu1999 replied:
其實我覺得這種問題最終會回到公司面來看，看公司或技術主管自己的 Coding Style 是怎樣。以這個專案來說你就是 Owner，理論上你會比我更熟這個專案，我希望你能評估後決定。

不過可以提供給你參考：

- 我之前的公司案子偏中小型，DAU 可能一百多這種，會拆 Service，Form Req 和 Repo 不會拆
- 現在公司是平台，DAU 約莫一萬吧我猜，會拆 Service，Form Req 看情況拆（我也還不清楚拆分標準），Repo 看情況拆，但一樣也沒有拆分的標準

JYu1999 replied:
比較純粹的要拆/不要拆這種二元式的決策，我更希望看到的是有一套明確的拆分標準。但倘若你覺得這沒有很重要，可以先從簡。

devusexu replied:
謝謝你分享案例
考量到 controller 的職責，我新增了兩個 service
form request 以及 repository 等到程式變得更複雜或有邏輯需要重複使用時再用

JYu1999 commented on routes/web.php:
```diff
@@ -16,3 +18,9 @@
 Route::get('/', function () {
     return view('welcome');
 });
```
沒有用到的可能要刪掉

devusexu committed:
refactor: remove $fillable, assign fields individually

devusexu committed:
refactor: remove unused route

devusexu committed:
refactor: move logic from controllers into dedicated services

1. remove RegisterController; use 'create' function in StudentController to register
2. add StudentService, RegistrationTokenService

devusexu committed:
refactor: remove unused route

devusexu committed:
feat: implement pop-up window for displaying student information

1. Added a 'info' button on the progress page to allow admins to view a student's information.
2. Introduced custom styles

devusexu committed:
fix: resolve incorrect progress evaluation

The issue arose from using format('%a'), which represents the interval as an absolute difference in days, leading to positive passed_days even when start_date is in the future. The problem is addressed by checking the relationship of the two dates first.

devusexu committed:
chore: remove 'isHoliday' column

JYu1999 commented on database/migrations/2024_02_28_212411_create_attendances_table.php:
```diff
@@ -0,0 +1,34 @@
+            $table->timestamps();
+            $table->foreignId('user_id')->constrained()->cascadeOnDelete();
+            $table->date('date');
+            $table->boolean('isHoliday')->default(0);
```
話說為什麼只有這個是 camel case 啊

devusexu replied:
沒注意到，已移除
